src/ndindex.jl

```diff
@@ -55,13 +55,17 @@ _flatten(i::Base.AbstractCartesianIndex) = _flatten(Tuple(i)...)
 @inline function _flatten(i::Base.AbstractCartesianIndex, I...)
     return (_flatten(Tuple(i)...)..., _flatten(I...)...)
  end
-Base.Tuple(index::NDIndex) = index.index
+Base.Tuple(@nospecialize(x::NDIndex)) = getfield(x, :index)
 
-Base.show(io::IO, i::NDIndex) = (print(io, "NDIndex"); show(io, Tuple(i)))
+Base.show(io::IO, @nospecialize(x::NDIndex)) = show(io, MIME"text/plain"(), x)
+function Base.show(io::IO, m::MIME"text/plain", @nospecialize(x::NDIndex))
+    print(io, "NDIndex")
+    show(io, m, Tuple(x))
+end
 
 # length
-Base.length(::NDIndex{N}) where {N} = N
-Base.length(::Type{NDIndex{N,I}}) where {N,I} = N
+Base.length(@nospecialize(x::NDIndex))::Int = length(Tuple(x))
+Base.length(@nospecialize(T::Type{<:NDIndex}))::Int = @inbounds(T.parameters[1])
```

Comment:
This change is incorrect. There is no guarantee about the order of parameters in a type that is <:NDIndex (or that they're even DataTypes at all:

```
julia> const NDIndex2{I<:Tuple{Vararg{Union{StaticInt,Int},2}}} = NDIndex{2, I}
NDIndex{2, I} where I<:Tuple{Union{Int64, StaticInt}, Union{Int64, StaticInt}}

julia> length(ans)
ERROR: type UnionAll has no field parameters
Stacktrace:
 [1] getproperty
   @ ./Base.jl:32 [inlined]
 [2] length(T::Type{<:NDIndex})
   @ Static ~/.julia/packages/Static/x0MGi/src/ndindex.jl:68
 [3] top-level scope
   @ REPL[3]:1
```

Reply:
> There is no guarantee about the order of parameters in a type that is <:NDIndex (or that they're even DataTypes at all:

I'm assuming we can still count on the order of parameters once we know it's a DataType thought right? So unwrapping UnionAll would work like:

```julia
function Base.length(@nospecialize(T::Type{<:NDIndex}))::Int
    if T isa UnionAll
        return length(T.body)
    else
        return @inbounds(T.parameters[1])
    end
end
```

Reply:
No, that's broken also. It could at the very least be a Union and even then you're relying on normalization rules that are not guaranteed to be stable. Just use the form it had before. That's the correct way to write it. If there's a performance issue with that, file a base issue.

Reply:
Fixed with #61